Frr-k8s dynamic log level

andreaskaris opened the pull request:
**Is this a BUG FIX or a FEATURE ?**:

/kind feature

**What this PR does / why we need it**:

This PR introduces the v1beta1 FRROperatorConfiguration Resource with support for configurable logging levels for the Operator and FRR. The key improvements include:

- **New FRROperatorConfiguration**: Adds a new CRD with configurable logging support for FRR instances and the operator, allowing fine-grained control over debug output without having to restart components

**Special notes for your reviewer**:

> **Note: **The bump to ... is needed because otherwise I get (after a revert):
```
[akaris@metallb e2etests]$ go test . -v -ginkgo.focus '.*logging.*' -ginkgo.v
# go.universe.tf/e2etest/pkg/metallb
../../../go/pkg/mod/github.com/metallb/metallb/e2etest@v0.0.0-20250127152015-3f9e99adbb0e/pkg/metallb/config.go:58:29: cannot use &metav1.Duration{…} (value of type *"k8s.io/apimachinery/pkg/apis/meta/v1".Duration) as "k8s.io/apimachinery/pkg/apis/meta/v1".Duration value in struct literal
../../../go/pkg/mod/github.com/metallb/metallb/e2etest@v0.0.0-20250127152015-3f9e99adbb0e/pkg/metallb/config.go:59:6: unknown field EnableGracefulRestart in struct literal of type "go.universe.tf/metallb/api/v1beta2".BGPPeerSpec
../../../go/pkg/mod/github.com/metallb/metallb/e2etest@v0.0.0-20250127152015-3f9e99adbb0e/pkg/metallb/config.go:91:17: peers[i].Spec.EnableGracefulRestart undefined (type "go.universe.tf/metallb/api/v1beta2".BGPPeerSpec has no field or method EnableGracefulRestart)
FAIL    github.com/metallb/frrk8stests [build failed]
FAIL
```
> I don't remember exactly any more what it was but it has to do with the updater.

-----------------

### Goals

- Provide a Kubernetes-native way to configure log levels at runtime via CRD
- Support dynamic log level changes for both FRR-K8s controllers and FRR daemon instances
- Maintain backward compatibility with existing `--log-level` command-line flags
- Ensure thread-safe log level updates across multiple controllers
- Allow per-instance configuration without requiring pod restarts

### Non-Goals

- Configuration of other runtime parameters beyond logging (e.g., FRR daemon options, BGP timers)
- Per-controller or per-daemon granular log level configuration (all components share the same log level)
- Log level persistence across operator reinstallations (configuration is stored in the CRD only)
- Log rotation or log aggregation features

### Design Details

#### New CRD: FRRK8sConfiguration

A new CRD `FRRK8sConfiguration` is introduced in the `frrk8s.metallb.io/v1beta1` API group:

```go
// FRRK8sConfigurationSpec defines the desired state of FRRK8sConfiguration.
type FRRK8sConfigurationSpec struct {
    // LogLevel sets the logging verbosity for the FRR-K8s components at runtime.
    // When configured, this value overrides the defaults established by the --log-level CLI flag.
    // Valid values are: all, debug, info, warn, error, none.
    // +kubebuilder:validation:Enum=all;debug;info;warn;error;none
    // +optional
    LogLevel string `json:"logLevel,omitempty"`
}
```

**Resource Requirements:**
- **Name**: The resource must be named `config`
- **Namespace**: Must reside in the namespace watched by FRR-K8s (typically `metallb-system`)
- **Scope**: namespace-scoped configuration, but only watched in frr-k8s namespace

**Example Resource:**

```yaml
apiVersion: frrk8s.metallb.io/v1beta1
kind: FRRK8sConfiguration
metadata:
  name: config
  namespace: metallb-system
spec:
  logLevel: debug
```

### Why a Dedicated Controller?

During the design phase, there was discussion about whether to create a dedicated `FRRK8sConfigurationController` or embed the logic within existing controllers. The final implementation uses a dedicated controller for the following reasons:

**Benefits of Dedicated Controller:**
- **Separation of Concerns**: Controller log level management is logically separate from FRR configuration management
- **Code Clarity**: Makes it immediately obvious where controller log levels are being set
- **Consistent Pattern**: Same controller can be instantiated in multiple binaries (`frr-k8s-controller`, `frr-status`, `statuscleaner`)
- **Future Extensibility**: Provides a natural place to add other runtime configuration options in the future

**Trade-offs:**
- **Additional Controller**: Adds one more controller reconciliation loop
- **Dual Watch**: Both `FRRK8sConfigurationController` and `FRRConfigurationController` watch the same CRD (though for different purposes)

**Release note**:
```release-note
Add FRROperatorConfiguration CR that allows users to configure logging at runtime without the need for restarts.

Default log levels are still set via the --log-level command line flag.

Users can now set logging levels for all components at runtime via FRROperatorConfiguration .spec.logLevel. 

The FRROperatorConfiguration object must be named `config` and must reside in the namespace that is watched by frr-k8s.
```

Fixes: https://github.com/metallb/frr-k8s/issues/376

andreaskaris commented on internal/logging/logging.go:
```diff
@@ -64,24 +65,50 @@ func Init(lvl string) (log.Logger, error) {
+
+// Log prints log output.
+func (dl *DynamicLvlLogger) Log(keyvals ...interface{}) error {
+	lvlLogger := level.NewFilter(dl.logger, dl.lvl)
```
the level.NewFilter can restrict log levels, but not set it.

So if I set (pseudo code), in that order:
```
l = level.NewFilter(l, "debug")
l = level.NewFilter(l, "error")
l = level.NewFilter(l, "info")
```
... then the log level will still be 'error', because that's the most restrictive.

So, I need a new struct that stores a logger (without the filter applied), and the desired log level. Then, right before logging, I apply the filter once

andreaskaris commented on api/v1beta1/frroperatorconfiguration_types.go:
```diff
@@ -0,0 +1,76 @@
+type FRROperatorConfigurationSpec struct {
+	// LogLevel defines the log level for the FRR-K8s controller and FRR daemons.
+	// +optional
+	LogLevel FRROperatorLogLevel `json:"logLevel,omitempty"`
```
will do

andreaskaris commented on internal/controller/frrconfiguration_controller.go:
```diff
@@ -273,3 +288,22 @@ func filterNodeEvent(e event.UpdateEvent, thisNode string) bool {
+		return "",
+			fmt.Errorf("more than a single configuration object found in namespace, objects: %+v",
+				configs.Items)
+	}
```
sounds good, I'll do that. what name though? just "config"?
(tbh I did it this way because kustomize's name prepending is a pita, but I'll change it)

andreaskaris replied:
using "config"

andreaskaris commented on internal/frr/frr_test.go:
```diff
@@ -939,6 +939,198 @@ func TestSingleUnnumberedSession(t *testing.T) {
 	testCheckConfigFile(t)
 }
 
+func TestLogLevelDebugging(t *testing.T) {
```
I had put that test in there to test:
```
	if config.Loglevel == "" {
		config.Loglevel = f.logLevel
	}
```

karampok commented:
I see that there are a number of commits that are not the feature itself

```
b58616e0 - Fix BGPConfig routers field to be omitempty
 c063b382 - Add checkgenerated target and use that in ci.yaml (general tooling improvement)
  285cd73c - Fix build make target (general fix)
  698f0a91 - E2E: bump metallb/e2etest and metallb dependency to v0.15.2 (dependency update)
```
Can we have them in a different PR ?

karampok commented on .github/workflows/ci.yaml:
```diff
@@ -50,22 +50,7 @@ jobs:
-
-      - name: Check if docs are different
-        run: make checkuncommitted
+          make checkgenerated
```
I  personally prefer to keep as is and not extend the Makefile

andreaskaris commented on internal/frr/frr.go:
```diff
@@ -52,7 +56,9 @@ func (f *FRR) ApplyConfig(config *Config) error {
-	config.Loglevel = f.logLevel
+	if config.Loglevel == "" {
+		config.Loglevel = f.logLevel
+	}
```
done

oribon commented on charts/frr-k8s/templates/frroperatorconfiguration.yaml:
ah is this the reason for the 1st commit? I think it's cleaner to specify the default loglevel through the flag in any case

andreaskaris replied:
I removed the file

oribon commented on internal/logging/logging.go:
```diff
@@ -53,8 +155,8 @@ func (l levelSlice) String() string {
 // mutates the contents of the flag package as well as os.Stderr.
-func Init(lvl string) (log.Logger, error) {
-	l := log.NewJSONLogger(log.NewSyncWriter(os.Stdout))
+func Init(w io.Writer, lvl *Level) (*DynamicLvlLogger, error) {
```
can `lvl` here not be a pointer?

oribon commented on cmd/status/controller/bgpsessionstate_controller.go:
```diff
@@ -78,13 +82,37 @@ func (ps peersStatus) remove(peer string) {
+				"err", err, "namespace", r.Namespace, "name", r.FRROperatorConfigurationName)
+		} else {
+			return ctrl.Result{}, err
+		}
```
the logging here is unnecessary imo, let's just return an err if it's not !isNotFound

andreaskaris replied:
done

oribon commented on cmd/status/controller/bgpsessionstate_controller.go:
```diff
@@ -142,13 +170,18 @@ func (r *BGPSessionStateReconciler) Reconcile(ctx context.Context, req ctrl.Requ
+	if r.DefaultLogLevel == nil {
+		return fmt.Errorf("BGPSessionStateReconciler SetupWithManager cannot start with nil DefaultLogLevel")
+	}
+
```
similar to other comments, I'd expect this to be verified from the caller

oribon commented on internal/controller/common_test.go:
is it intentional that we didn't update frrstate_controller with the new logger?
also, can we add a test here to see that the controllers react properly to changes in the config resource?

andreaskaris commented on cmd/status/controller/bgpsessionstate_controller.go:
```diff
@@ -78,13 +82,37 @@ func (ps peersStatus) remove(peer string) {
+	if err = r.Logger.SetLogLevel(currentLogLevel); err != nil {
+		return ctrl.Result{}, err
+	}
+	level.Info(r.Logger).Log("controller", "BGPSessionState", "log level controller", currentLogLevel.String())
```
done

karampok commented on e2etests/pkg/config/config.go:
```diff
@@ -68,8 +70,49 @@ func (u *Updater) Update(secrets []v1.Secret, configs ...frrk8sv1beta1.FRRConfig
 	return nil
 }
 
+// UpdateFrrOperatorConfiguration creates or updates the given FRROperatorConfiguration.
```
UpdateFRR

andreaskaris replied:
The method is called UpdateFRROperatorConfiguration because it creates or updates an FRROperatorConfiguration, and the comment of the public function starts with the function name, so I think that's ok as is. The reason not to put it into the Update method was that I didn't want to pollute the existing stuff / add convoluted parameters to it

karampok commented on e2etests/tests/logging.go:
```diff
@@ -0,0 +1,658 @@
+	"go.universe.tf/e2etest/pkg/metallb"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
```
imported twice

karampok commented on e2etests/tests/logging.go:
```diff
@@ -0,0 +1,658 @@
+	defer ginkgo.GinkgoRecover()
+	updater, err := config.NewUpdater()
+	Expect(err).NotTo(HaveOccurred())
+	reporter := dump.NewK8sReporter(k8s.FRRK8sNamespace)
```
All these should be inside before each, 

https://onsi.github.io/ginkgo/#no-assertions-in-container-nodes

 Why we need GInkgoRecover?

andreaskaris replied:
done in next push

karampok commented on e2etests/pkg/k8s/namespace.go:
```diff
@@ -3,6 +3,7 @@
-	FRRK8sDaemonset = "frr-k8s-daemon"
+	FRRK8sNamespace        = "frr-k8s-system"
+	FRRK8sDaemonsetLS      = "app.kubernetes.io/component=frr-k8s"
+	FRRK8sStatusCleanerApp = "app.kubernetes.io/component=statuscleaner"
```
should we have here
```
frrK8SContainerName              = "controller"
	frrK8SStatusContainerName        = "frr-status"
	frrK8SStatusCleanerContainerName = "frr-k8s-statuscleaner"
	frrContainerName                 = "frr"
	frrOperatorConfigurationName     = "config"
```
?

karampok commented on e2etests/tests/logging.go:
```diff
@@ -0,0 +1,658 @@
+// change at all (if updated to FRRConfiguration happen quickly in a row). It's also not strictly necessary to verify
+// that the logging of the FRRState controller works correctly, as all the update logic is driven via the
+// FRRConfigurationReconciler and therefore the FRRState reconciler's log level changes will work exactly the same.
+func expectedLogsByLogLevel(logLevel string) types.GomegaMatcher {
```
It would first time ( I think) returninig types.GomegaMatcher, I would find that pattern.

Can we have simpler match like just find `level:debug`

karampok commented on e2etests/tests/logging.go:
```diff
@@ -0,0 +1,658 @@
+	return Not(ContainSubstring(`"controller":"NodeStateCleaner","level":"debug","log level controller":"%s"`, logLevel))
+}
+
+// checkLogs updates the FRR configuration and verifies that controller logs match expected patterns.
```
Does many stuff

andreaskaris replied:
But that's a question of preference, right?
The reason to have it that way is that I don't have to do the same code in the 5 locations where it's being called, and to be able to just have a one-liner there.
I had it all explicitly without that function, but when I saw how much redundancy there was, I decided to move it into this function.
I agree in principal on return string / bool / error and then doing ginkgo assertions, but here I'm testing with Eventually inside the function, and that returns AsyncAssertion, and not an error, and goes directly together with the Should

karampok commented on e2etests/tests/logging.go:
```diff
@@ -0,0 +1,658 @@
+	frrContainerName                 = "frr"
+	frrOperatorConfigurationName     = "config"
+
+	dummyRawConfig = `router bgp 64512
```
do we really need dummy config?

andreaskaris replied:
Good catch, no, it's not needed.

Removed in next push

andreaskaris commented on e2etests/tests/logging.go:
```diff
@@ -0,0 +1,658 @@
+        neighbor 172.18.0.5 port 179`
+)
+
+var _ = ginkgo.Describe("Verifyng dynamic logging levels for Operator", func() {
```
I made it more succinct, reorganized the tests a bit, and dropped one that was redundant -  changed to:
```
Logging for status cleaner when FRROperatorConfiguration changes logs with the correct log levels
Logging for frr-k8s when no FRROperatorConfiguration is present logs with default log level
Logging for frr-k8s when FRROperatorConfiguration with empty spec is added logs with default log level
Logging for frr-k8s when FRRConfiguration is added logs with info level when logLevel is set to info
Logging for frr-k8s when FRRConfiguration is added logs with error level when logLevel is set to error
Logging for frr-k8s when FRROperatorConfiguration is removed logs with default log level
Logging for frr when no logLevel configuration is present logs with default log level
Logging for frr when empty logLevel configuration is present logs with default log level
Logging for frr when logLevel configuration with logLevel is present logs with configured log level
```

done in the next push

andreaskaris commented on e2etests/tests/logging.go:
```diff
@@ -0,0 +1,658 @@
+	return ""
+}
+
+// getDefaultLogLevel retrieves the default log level from the FRR DaemonSet for the specified container.
```
The log levels are different between helm manifests and the deploy manifests.
I'd prefer determining it in code, given that it's already written

karampok replied:
in the past I was either tried to align (== same log level to all deployment ways) or make the ci yaml file to input the the status.  That was my approach.

fedepaol commented on internal/logging/logging.go:
```diff
@@ -22,15 +23,100 @@ import (
+
+// NewLevel parses the provided string and returns a pointer to a Level, or an error if the provided
+// level string is invalid.
+func NewLevel(l string) (Level, error) {
```
Let's put the logger definition first, the level afterwards. It's how a reader would read the file (from most important to less important)

fedepaol replied:
or even better, let's put it in a level.go file

andreaskaris replied:
fixed on next push in "Move levels to level.go"

fedepaol commented on internal/logging/logging.go:
```diff
@@ -64,24 +150,54 @@ func Init(lvl string) (log.Logger, error) {
+	return dl, nil
+}
+
+// DynamicLvlLogger is a logger with a log level that can be set at runtime.
```
no need to explain why we need a new logger type here (more suitable in the commit message).
Let's call this just "Logger" as for this project this is the logger type.

andreaskaris replied:
fixed in Rename NewDynamicLvlLogger to Logger on next push

fedepaol commented on internal/logging/logging.go:
```diff
@@ -64,24 +150,54 @@ func Init(lvl string) (log.Logger, error) {
+// NewDynamicLvlLogger creates a new DynamicLvlLogger that wraps the provided logger with
+// a dynamically adjustable log level filter. The initial log level is set according to the
+// provided level parameter.
+func NewDynamicLvlLogger(logger log.Logger, level Level) *DynamicLvlLogger {
```
no need for a constructor here imo

fedepaol commented:
Need to go over the e2e tests yet.

One comment on the overall architecture: all the controllers inherit the same logger instance (passed as pointer).
Now, having a side effect "somewhere else" because I change the loglevel of my local logger is not ideal / super easy to follow. Since we are using a shared logger _anyway, just dressed as a local logger_ , how about having it as a singleton of the logging package? If we do that, we could have one single controller listening for the new CR, calling a logging.SetLevel() function, that clearly changes the general log level.

andreaskaris commented on internal/controller/nodestate_cleaner_controller.go:
```diff
@@ -17,26 +19,45 @@ import (
+		return ctrl.Result{}, err
+	}
+
+	// Parse the current log level once here. Either, the FrrOperatorConfiguration log level is invalid (the empty
```
done

andreaskaris commented on api/v1beta1/frroperatorconfiguration_types.go:
```diff
@@ -0,0 +1,62 @@
+//nolint
+//+genclient
+
+// FRROperatorConfiguration holds the FRR Operator configuration with global
```
will do over the wknd

andreaskaris replied:
done

andreaskaris commented on internal/logging/logging.go:
```diff
@@ -64,24 +150,54 @@ func Init(lvl string) (log.Logger, error) {
+
+// Log prints log output.
+func (dl *DynamicLvlLogger) Log(keyvals ...interface{}) error {
+	lvlLogger := level.NewFilter(dl.logger, dl.lvl)
```
done

andreaskaris commented:
I addressed Federico's change requests -I already rebased  the rename operation and the change can be seen in:
https://github.com/metallb/frr-k8s/compare/af5288f953c8aed4e335b53d96478b6583225ab9..aca1ed3e31cd90ae358cd3b0d8797fa0d513bc4b

All other changes are inside dedicated commits (to make the review operation easier)

Once I got the o.k., I'll squash things down

fedepaol commented:
> > now that we have the logger as a singleton, what prevents us to have an frrk8sconfiguration (or whatever the name is) controller that reconciles and sets the level on just the singleton?
> 
> See a bit above: this approach would work perfectly fine if we had to set only the singleton's log level. However, we also have to set the FRR log level (write to file + reload), and thus the frrconfiguration controller will have to watch and reconcile frrk8sconfiguration objects. Thus, things here would be redundant
> 

not necessarily, because we can make the frrconfiguration controller do its business logic only (which is only translating the log level to the corresponding value in the frr config) while the frrk8sconfiguration reconciler would set the singleton's loglevel. 

> As mentioned earlier, I had implemented it the other way before, and actually wouldn't mind having it like that, meaning having an frrk8scontroller added to each binary for setting the controller log level; and living with the fact that the frrconfiguration controller will run an additional watch on the frrk8sconfguration resources to set the FRR log level - it's happening from the informers cache anyway, so I imagine the overhead is minimal.
> 
> The alternative is what we currently have, which also makes sense, if we don't want that extra controller. After all, it's prob just a question of preference.

I think I prefer the new controller approach, as in it's way more straightforward to follow. Let's focus on the complexity and maintenability, as the configuration will change rarely. What were the objections to having a single controller?

andreaskaris commented on internal/logging/logging.go:
```diff
@@ -21,67 +23,84 @@ import (
-func Init(lvl string) (log.Logger, error) {
-	l := log.NewJSONLogger(log.NewSyncWriter(os.Stdout))
+// The default log level is Info.
+func NewLogger(w io.Writer) (*Logger, error) {
```
done (next push)

andreaskaris commented on cmd/frr-k8s-controller/main.go:
```diff
@@ -89,11 +88,19 @@ func main() {
+		os.Exit(1)
+	}
+
+	if err := logging.Init(); err != nil {
```
Init() uses `once.Do`. If we add a parameter here, that becomes confusing to me. Could potentially call GetLogger().SetLogLevel(lvl) inside init itself, after the once.Do ... not sure if I find this more confusing, though
There's no more function parseloglevel

andreaskaris commented on internal/controller/frrconfiguration_controller.go:
```diff
@@ -97,13 +101,21 @@ func (r *FRRConfigurationReconciler) Reconcile(ctx context.Context, req ctrl.Req
+
+	// logLevelForApply is used exclusively to set the FRR instance's log level. The log level for the controllers is
+	// set via the FRRK8sConfiguration controller.
+	logLevelForApply, err := getLogLevel(ctx, r, r.Namespace, r.FRRK8sConfigurationName, r.DefaultLogLevel)
```
The simplification here doesn't work:

a) inside frrk8sconfiguration_controller.go:
if config.Spec.LogLevel == "" --> need to fall back to default as configured by `--log-level` flag (or its default info), so the defaultLogLevel parameter is needed here, cannot simply read the current GetLogLevel() as we might want to change that to default that's established at startup
b) insde frrconfiguration_controller.go:
We can be in a potential race here, where the frrconfiguration_controller is triggered before the frrk8sconfiguration_controller could even set the singleton's log level. Hence, the frrconfiguration controller needs to make the log level decision by itself, and that sends us back to the same scenario as in a)

Race condition example for b):

0. default log level info
1. current log-level error
2. log-level changed to "" 
3. frrconfiguration_controller runs first, cannot parse "", hence falls back. If it now reads the singleton's LogLevel, it will set the FRR configuration still to 'error' , but it should have set 'info'
4. frrk8sconfiguration_controller runs second, and sets the singleton's log level to 'info'

With the current code, that won't happen (other than the log messages for frrconfiguration controller for this one reconciliation still being at the other log level, but I wouldn't consider this a bug.

andreaskaris replied:
> you can still call this loglevel
done

fedepaol replied:
> The simplification here doesn't work:
> 
> a) inside frrk8sconfiguration_controller.go: if config.Spec.LogLevel == "" --> need to fall back to default as configured by `--log-level` flag (or its default info), so the defaultLogLevel parameter is needed here, cannot simply read the current GetLogLevel() as we might want to change that to default that's established at startup

right, and even if we make the loglevel mandatory (which I think it's fair) we'd still need to handle the case where
we remove the frrk8sconfiguration instance.

>  b) insde frrconfiguration_controller.go: We can be in a potential race here, where the frrconfiguration_controller is triggered before the frrk8sconfiguration_controller could even set the singleton's log level. Hence, the frrconfiguration controller needs to make the log level decision by itself, and that sends us back to the same scenario as in a)
> 
> Race condition example for b):
> 
>     0. default log level info
> 
>     1. current log-level error
> 
>     2. log-level changed to ""
> 
>     3. frrconfiguration_controller runs first, cannot parse "", hence falls back. If it now reads the singleton's LogLevel, it will set the FRR configuration still to 'error' , but it should have set 'info'
> 
>     4. frrk8sconfiguration_controller runs second, and sets the singleton's log level to 'info'
> 
> 
> With the current code, that won't happen (other than the log messages for frrconfiguration controller for this one reconciliation still being at the other log level, but I wouldn't consider this a bug.

andreaskaris commented on internal/controller/frrk8sconfiguration_controller.go:
```diff
@@ -0,0 +1,91 @@
+	defaultLogLevel logging.Level) (logging.Level, error) {
+	config := frrk8sv1beta1.FRRK8sConfiguration{}
+	err := r.Get(ctx, types.NamespacedName{Namespace: namespace, Name: frrK8sConfigurationName}, &config)
+	if err != nil && !k8serrors.IsNotFound(err) {
```
done

andreaskaris commented on internal/frr/level.go:
```diff
@@ -0,0 +1,41 @@
+
+// NewLevel converts the logging.Level to its corresponding FRR daemon log level.
+// Falls back to LevelInformational if the level is anything unexpected.
+func NewLevel(l logging.Level) Level {
```
done

andreaskaris commented on e2etests/tests/logging.go:
```diff
@@ -0,0 +1,589 @@
+				Expect(err).NotTo(HaveOccurred())
+
+				// Expect to find default log level configuration in FRR config.
+				expectedLogString := "log stdout"
```
```
type FRRConfigurationStatus struct {
	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
	// Important: Run "make" to regenerate code after modifying this file
}
```

I suppose that you mean FRRNodeState.status?
```
- apiVersion: frrk8s.metallb.io/v1beta1
  kind: FRRNodeState
  metadata:
    creationTimestamp: "2026-01-28T14:31:23Z"
    generation: 1
    name: kind-control-plane
    resourceVersion: "1107"
    uid: f70def5d-ec57-4f1b-824f-b8d4215f6a98
  spec: {}
  status:
    lastConversionResult: success
    lastReloadResult: success
    runningConfig: |
      Building configuration...

      Current configuration:
      !
      frr version 10.4.1_git
      frr defaults traditional
      hostname kind-control-plane
      log stdout informational
(..)
```

Unless I'm missing some other place where this could be stored, I think getting it from vtysh directly or via the frrnodestate is +/- the same. If I want to get it from the frrnodestate, I have to add code to init a new client with the correct scheme, and do essentially the same parsing

andreaskaris commented on e2etests/tests/logging.go:
```diff
@@ -0,0 +1,658 @@
+func checkLogs(cs clientset.Interface, updater *config.Updater, config frrk8sv1beta1.FRRConfiguration,
+	expected, expectedBGPSessionState types.GomegaMatcher) {
+	// We need to wait 1 second for PodLogsSinceTime not to include previous output (as granularity is 1s).
+	time.Sleep(1 * time.Second)
```
done but added comments and moved into a function to explain why

fedepaol commented:
LGTM! Thanks a lot!

karampok commented on e2etests/tests/logging.go:
```diff
@@ -0,0 +1,433 @@
+	})
+
+	ginkgo.Context("for status cleaner", func() {
+		var frrk8sStatusCleanerPods []*corev1.Pod
```
how many pods do we have? only one no?

andreaskaris replied:
yes - but I feel that this is safer in the future instead of picking pod[0] given that the status cleaner is a deployment

karampok commented on e2etests/tests/logging.go:
```diff
@@ -0,0 +1,433 @@
+			})
+		})
+
+		ginkgo.DescribeTable("when FRRConfiguration is added", func(logLevel string) {
```
the when here is FRRCOnfiguration is added, the when above is FRK8sConfiguration is present, I think ti should be aligned

andreaskaris commented on e2etests/tests/logging.go:
```diff
@@ -0,0 +1,433 @@
+				}
+				err := updater.UpdateFRRK8sConfiguration(operatorConfig)
+				Expect(err).NotTo(HaveOccurred())
+				sleepForMetav1Granularity()
```
I created a follow-up PR (currently WIP) for some of your points: https://github.com/metallb/frr-k8s/pull/396
